src/lib/alloc.c

```diff
@@ -555,36 +555,65 @@ static void *rmalloc_runtime(int zone, uint32_t caps, size_t bytes)
 	return get_ptr_from_heap(heap, zone, caps, bytes);
 }
 
+static void *rballoc(int zone, uint32_t caps, size_t bytes)
+{
+	void *ptr;
+
+	ptr = _balloc(zone, caps, bytes);
+
+#if CONFIG_DEBUG_HEAP
+	if (!ptr) {
+		trace_error(TRACE_CLASS_MEM,
+			    "failed to alloc 0x%x bytes caps 0x%x",
+			    bytes, caps);
+
+		alloc_trace_buffer_heap(zone, caps, bytes);
+	}
+#endif
+
+	return ptr;
+}
+
 /* allocates memory - not for direct use, clients use rmalloc() */
 void *_malloc(int zone, uint32_t caps, size_t bytes)
 {
 	uint32_t flags;
 	void *ptr = NULL;
 
+	if (bytes >= HEAP_BUFFER_BLOCK_SIZE)
+		return rballoc(zone, caps, bytes);
+
 	spin_lock_irq(&memmap.lock, flags);
 
 	switch (zone & RZONE_TYPE_MASK) {
 	case RZONE_SYS:
 		ptr = rmalloc_sys(zone, caps, cpu_get_id(), bytes);
 		break;
 	case RZONE_SYS_RUNTIME:
-		ptr = rmalloc_sys_runtime(zone, caps, cpu_get_id(), bytes);
+		ptr = rmalloc_sys_runtime(zone, caps, cpu_get_id(),
+					  bytes);
 		break;
 	case RZONE_RUNTIME:
 		ptr = rmalloc_runtime(zone, caps, bytes);
 		break;
 	default:
 		trace_mem_error("rmalloc() error: invalid zone");
-		panic(SOF_IPC_PANIC_MEM); /* logic non recoverable problem */
+
+		/* logic non recoverable problem */
+		panic(SOF_IPC_PANIC_MEM);
 		break;
 	}
 
+	spin_unlock_irq(&memmap.lock, flags);
```

Comment:
Here you are unlocking before another try to alloc.
It can possibly create race condition.